File: cs/src/AutoDiff/TermPower.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace AutoDiff
{
    public class TermPower : Term
    {
        public TermPower(Term baseTerm, Term exponent)
        {
            Base = baseTerm;
            Exponent = exponent;
        }

        /// <summary>
        /// Gets the base term of the power function
        /// </summary>
        public Term Base { get; private set; }

        /// <summary>
        /// Gets the exponent term of the power function.
        /// </summary>
        public Term Exponent { get; private set; }

        /// <summary>
        /// Accepts a term visitor.
        /// </summary>
        /// <param name="visitor">The term visitor to accept.</param>
        public override void Accept(ITermVisitor visitor)
        {
            visitor.Visit(this);
        }

        /// <summary>
        /// Accepts a term visitor with a generic result
        /// </summary>
        /// <typeparam name="TResult">The type of the result from the visitor's function</typeparam>
        /// <param name="visitor">The visitor to accept</param>
        /// <returns>
        /// The result from the visitor's visit function.
        /// </returns>
        public override TResult Accept<TResult>(ITermVisitor<TResult> visitor)
        {
            return visitor.Visit(this);
        }
		public override string ToString()
		{
			return string.Format("termPower( {0}, {1} )", Base, Exponent);
		}
		
		public override Term AggregateConstants ()
		{
			Base = Base.AggregateConstants();
			Exponent = Exponent.AggregateConstants();
			if (Exponent is Zero) {
				return 1;
			}
			if (Base is Constant && Exponent is Constant) {
				return Math.Pow ((Base as Constant).Value, (Exponent as Constant).Value);
			} else if (Base is Zero) {				
				return Base;
			} else if (Base is TermPower) {
				Exponent *= (Base as TermPower).Exponent;
				Base = (Base as TermPower).Base;
				return this;
			  } else {
				return this;
			}
		}
		public override Term Derivative(Variable v)
		{
			return new TermPower(this.Base,this.Exponent-1)*(this.Exponent*this.Base.Derivative(v)+this.Base*(new Log(this.Base))*this.Exponent.Derivative(v));
		}
    }
}

File: java/src/main/java/AutoDiff/TermPower.java
//
// Translated by CS2J (http://www.cs2j.com): 12.11.2016 09:40:34
//

package AutoDiff;

public class TermPower  extends Term
{
    public TermPower(Term baseTerm, Term exponent) throws Exception {
        setBase(baseTerm);
        setExponent(exponent);
    }

    /**
    * Gets the base term of the power function
    */
    private Term Base;
    public Term getBase() {
        return Base;
    }

    public void setBase(Term value) {
        Base = value;
    }

    /**
    * Gets the exponent term of the power function.
    */
    private Term Exponent;
    public Term getExponent() {
        return Exponent;
    }

    public void setExponent(Term value) {
        Exponent = value;
    }

//    /**
//    * Accepts a term visitor.
//    *
//    *  @param visitor The term visitor to accept.
//    */
//    public void accept(ITermVisitor visitor) throws Exception {
//        visitor.visit(this);
//    }

    /**
    * Accepts a term visitor with a generic result
    * The type of the result from the visitor's function
    *  @param visitor The visitor to accept
    *  @return 
    * The result from the visitor's visit function.
    */
    public <TResult>TResult accept(ITermVisitor<TResult> visitor) throws Exception {
        return visitor.visit(this);
    }

    public String toString() {
        try
        {
            return String.format("termPower( %s, %s )", getBase(), getExponent());
        }
        catch (RuntimeException exception)
        {
            throw exception;
        }
        catch (Exception exception)
        {
            throw new RuntimeException(exception);
        }
    
    }

    public Term aggregateConstants() throws Exception {
        setBase(getBase().aggregateConstants());
        setExponent(getExponent().aggregateConstants());
        if (getExponent() instanceof Zero)
        {
            return TermBuilder.Constant(1);
        }
         
        if (getBase() instanceof Constant && getExponent() instanceof Constant)
        {
            return TermBuilder.Constant(Math.pow((getBase() instanceof Constant ? (Constant)getBase() : (Constant)null)
                    .getValue(),
                (getExponent() instanceof Constant ? (Constant)getExponent() : (Constant)null).getValue()));
        }
        else if (getBase() instanceof Zero)
        {
            return getBase();
        }
        else if (getBase() instanceof TermPower)
        {
            setExponent(multiply(getExponent(), (getBase() instanceof TermPower ? (TermPower)getBase() :
                (TermPower)null).getExponent()));
            setBase((getBase() instanceof TermPower ? (TermPower)getBase() : (TermPower)null).getBase());
            return this;
        }
        else
        {
            return this;
        }   
    }

    public Term derivative(Variable v) throws Exception {
        return new TermPower(this.getBase(), multiply(substract(this.getExponent(), TermBuilder.Constant(1)),
            add(multiply(this.getExponent(), this.getBase().derivative(v)), multiply(multiply(this
                    .getBase(), (new Log(this.getBase()))), this.getExponent().derivative(v)))));
    }

}


